src/procedures/proc_common_neighbors.c

```diff
@@ -0,0 +1,316 @@
+
+	info = GxB_Scalar_setElement_INT64(ctx->thunk, min);
+	assert(info == GrB_SUCCESS);
+
+	info = GxB_Matrix_select(ctx->SIM, GrB_NULL, GrB_NULL, ctx->select_op, ctx->SIM, ctx->thunk,
+							 ctx->desc);
+	assert(info == GrB_SUCCESS);
+
+	info = GrB_Matrix_nvals(&nvals, ctx->SIM);
+	assert(info == GrB_SUCCESS);
+
+	// Force flush.
+	GrB_Matrix_nvals(&nvals, ctx->SIM);
+
+	// Reset iterator.
+	if(!ctx->iter) GxB_MatrixTupleIter_new(&ctx->iter, ctx->SIM);
+	else GxB_MatrixTupleIter_reuse(ctx->iter, ctx->SIM);
+}
+
+ProcedureResult Proc_CommonNeighborsInvoke(ProcedureCtx *ctx, const SIValue *args) {
+	if(array_len((SIValue *)args) != 1) return PROCEDURE_ERR;
+	if(!(SI_TYPE(args[0]) & T_STRING)) return PROCEDURE_ERR;
+
+	const char *relation = args[0].stringval;
+
+	GrB_Info info;
+	GrB_Index nvals;
+	GrB_Index nrows;
+	GrB_Index ncols;
+	GrB_Index batch_size;
+	GrB_Matrix A = GrB_NULL;
+	GrB_Matrix At = GrB_NULL;
+	GrB_Matrix SIM = GrB_NULL;
+	GxB_Scalar thunk = GrB_NULL;
+	GrB_Vector DEGREE = GrB_NULL;
+	GrB_Descriptor desc = GrB_NULL;
+	GxB_SelectOp select_op = GrB_NULL;
+
+	Graph *g = QueryCtx_GetGraph();
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+
+	Schema *s = GraphContext_GetSchema(gc, relation, SCHEMA_EDGE);
+	if(!s) return PROCEDURE_OK;
+	A = Graph_GetRelationMatrix(g, s->id);
+
+	info = GrB_Matrix_nrows(&nrows, A);
+	assert(info == GrB_SUCCESS);
+
+	info = GrB_Matrix_ncols(&ncols, A);
+	assert(info == GrB_SUCCESS);
+
+	batch_size = MIN(nrows, BATCH_SIZE);
+
+	// Create similarity matrix.
+	info = GrB_Matrix_new(&SIM, GrB_INT64, batch_size, ncols);
+	assert(info == GrB_SUCCESS);
+
+	// Create transposed relationship matrix.
+	info = GrB_Matrix_new(&At, GrB_BOOL, nrows, ncols);
+	assert(info == GrB_SUCCESS);
+
+	// Transpose relationship matrix.
+	info = GrB_transpose(At, GrB_NULL, GrB_NULL, A, GrB_NULL);
+	assert(info == GrB_SUCCESS);
+
+	// Create degree vector, DEGREE[i] = out degree of node i.
+	info = GrB_Vector_new(&DEGREE, GrB_INT64, ncols);
+	assert(info == GrB_SUCCESS);
+
+	// Reduce A rows, using plus to sum each row.
+	info = GrB_Matrix_reduce_Monoid(DEGREE, GrB_NULL, GrB_NULL, GxB_PLUS_INT64_MONOID, A, GrB_NULL);
+	assert(info == GrB_SUCCESS);
+
+	info = GxB_Scalar_new(&thunk, GrB_INT64);
+	assert(info == GrB_SUCCESS);
+
+	info = GxB_SelectOp_new(&select_op, _select_triu, GrB_INT64, GrB_INT64);
+	assert(info == GrB_SUCCESS);
+
+	info = GrB_Descriptor_new(&desc);
+	assert(info == GrB_SUCCESS);
+	GrB_Descriptor_set(desc, GrB_OUTP, GrB_REPLACE);
+
+	// Update procedure context.
+	SimilarityContext *sim_ctx = SimilarityContext_New();
+	sim_ctx->A = A;
+	sim_ctx->At = At;
+	sim_ctx->SIM = SIM;
+	sim_ctx->desc = desc;
+	sim_ctx->nrows = nrows;
+	sim_ctx->thunk = thunk;
+	sim_ctx->select_op = select_op;
+	sim_ctx->batch_size = batch_size;
+	sim_ctx->DEGREE = DEGREE;
+
+	ctx->privateData = sim_ctx;
+	return PROCEDURE_OK;
+}
+
+SIValue *Proc_CommonNeighborsStep(ProcedureCtx *ctx) {
+	assert(ctx->privateData);
```

Comment:
We should return NULL here instead of asserting, as this will trigger a crash if a non-existent edge is specified.